google/cloud/firestore_bundle/bundle.py

```diff
@@ -0,0 +1,439 @@
+# limitations under the License.
+
+"""Classes for representing bundles for the Google Cloud Firestore API."""
+
+import datetime
+import json
+
+from google.cloud.firestore_bundle.types.bundle import (
+    BundledDocumentMetadata,
+    BundledQuery,
+    BundleElement,
+    BundleMetadata,
+    NamedQuery,
+)
+from google.cloud._helpers import _datetime_to_pb_timestamp, UTC  # type: ignore
+from google.cloud.firestore_bundle._helpers import limit_type_of_query
+from google.cloud.firestore_v1.async_query import AsyncQuery
+from google.cloud.firestore_v1.base_client import BaseClient
+from google.cloud.firestore_v1.base_document import DocumentSnapshot
+from google.cloud.firestore_v1.base_query import BaseQuery
+from google.cloud.firestore_v1.document import DocumentReference
+from google.cloud.firestore_v1.query import Query
+from google.cloud.firestore_v1 import _helpers
+from google.protobuf.timestamp_pb2 import Timestamp  # type: ignore
+from typing import (
+    Dict,
+    Generator,
+    List,
+    Optional,
+    Union,
+)
+
+
+class FirestoreBundle:
+    """A group of serialized documents and queries, suitable for
+    longterm storage.
+
+    If any queries are added to this bundle, all associated documents will be
+    loaded and stored in memory for serialization.
+
+    Usage:
+
+        from google.cloud.firestore import Client
+        from google.cloud.firestore_bundle import FirestoreBundle
+        from google.cloud.firestore import _helpers
+
+        db = Client()
+        bundle = FirestoreBundle('my-bundle')
+        bundle.add('all-users', db.collection('users')._query())
+        bundle.add(
+            'top-ten-hamburgers',
+            db.collection('hamburgers').limit(limit=10)._query(),
+        )
+        serialized: str = bundle.build()
+
+        # Store somewhere like your GCS or your device's file system
+
+        # Some time later after loading `serialized` again
+        bundle = _helpers.deserialize_bundle(serialized)
+
+        users: Iterable[DocumentSnapshot] = bundle.get_documents(
+            query_name='all-users',
+        )
+
+    Args:
+        name (str): The Id of the bundle.
+    """
+
+    BUNDLE_SCHEMA_VERSION: int = 1
+
+    def __init__(self, name: str) -> None:
+        self.name: str = name
+        self.documents: Dict[str, "_BundledDocument"] = {}
+        self.named_queries: Dict[str, NamedQuery] = {}
+        self.latest_read_time: Timestamp = Timestamp(seconds=0, nanos=0)
+        self._deserialized_metadata: Optional[BundledDocumentMetadata] = None
+
+    def add(
+        self,
+        document_or_query_name: Union[DocumentSnapshot, str],
+        query: Optional[Query] = None,
+    ) -> "FirestoreBundle":
+        """Adds a document or entire query to the bundle.
+
+        This function is offered for convenience / API parity with other SDKs.
+        Most users will get better outcomes from using the more specific,
+        `add_document` or `add_named_query` methods.
```

Comment:
I think this function is confusing and we can drop it for the better alternatives.

Reply:
+1

Reply:
I'm happy to drop this - I'd only reluctantly added it for parity (as the docstring says). But if you two don't think it's necessary, I am happy to nuke it.